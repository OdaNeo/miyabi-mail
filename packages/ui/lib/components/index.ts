export * from './button_temp';
export * from './input';
export * from './label';
export * from './textarea';
export * from './select';
export * from './switch';
export * from './popover';
export * from './tooltip';
export * from './progress';
export * from './slider';
