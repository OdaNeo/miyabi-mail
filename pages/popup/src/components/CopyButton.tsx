import { Button } from '@extension/ui';
import { motion, AnimatePresence } from 'framer-motion';
import { Copy, Check } from 'lucide-react';
import { useState } from 'react';

export const CopyButton = ({ handleCopyText }: { handleCopyText: () => void }) => {
  const [isCopied, setIsCopied] = useState(false);

  const handleCopy = () => {
    if (isCopied) {
      return;
    }
    handleCopyText();
    setIsCopied(true);
    setTimeout(() => setIsCopied(false), 2000);
  };

  return (
    <Button
      variant="ghost"
      size="icon"
      data-testid="copy-icon-group"
      className="h-6 w-6 text-slate-500 hover:text-slate-900 relative"
      onClick={handleCopy}
    >
      <AnimatePresence initial={false} mode="wait">
        {isCopied ? (
          <motion.div
            key="check"
            initial={{ opacity: 0, scale: 0.8 }}
            animate={{ opacity: 1, scale: 1 }}
            exit={{ opacity: 0, scale: 0.8 }}
            transition={{ duration: 0.2 }}
            className="absolute inset-0 flex items-center justify-center"
          >
            <Check className="h-4 w-4 text-green-500" />
          </motion.div>
        ) : (
          <motion.div
            key="copy"
            initial={{ opacity: 0, scale: 0.8 }}
            animate={{ opacity: 1, scale: 1 }}
            exit={{ opacity: 0, scale: 0.8 }}
            transition={{ duration: 0.2 }}
            className="absolute inset-0 flex items-center justify-center"
          >
            <Copy className="h-4 w-4" />
          </motion.div>
        )}
      </AnimatePresence>
    </Button>
  );
};
