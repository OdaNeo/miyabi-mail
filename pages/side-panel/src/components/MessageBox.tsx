import clsx from 'clsx';
import { twMerge } from 'tailwind-merge';
import type React from 'react';

export const MessageBox: React.FC<{ children: React.ReactNode }> = ({ children }) => {
  return (
    <div
      className={twMerge(
        clsx(`h-auto overflow-y-auto text-sm scrollbar-custom
           dark:bg-slate-800 dark:text-slate-100 dark:placeholder-slate-400 
         bg-white text-slate-900 placeholder-slate-500 ${children && `p-3`}`),
      )}
    >
      {children}
    </div>
  );
};
