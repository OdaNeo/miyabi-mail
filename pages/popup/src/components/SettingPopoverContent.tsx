import { Button } from '@src/components/ui/button';
import { AlertTriangle, EyeOff, Eye, Trash2 } from 'lucide-react';
import { useRef, useState } from 'react';
import { Input } from '@src/components/ui/input';
import { Select, SelectContent, SelectItem, SelectTrigger, SelectValue } from './ui/select';
import { useStorage } from '@extension/shared';
import { apiKeyStorage, apiVersionStorage } from '@extension/storage';
import type React from 'react';
import { useI18n } from '@src/hooks/useI18n';
import { CopyButton } from './CopyButton';

export const SettingPopoverContent = ({ setIsOpen }: { setIsOpen: (isOpen: boolean) => void }) => {
  const [showKey, setShowKey] = useState(false);

  const apiVersion = useStorage(apiVersionStorage);
  const apiKey = useStorage(apiKeyStorage);

  const inputRef = useRef<HTMLInputElement>(null);

  const { API_KEY_SETTING, PLEASE_SET_API_KEY, SAVE } = useI18n();

  const handleClosePopover = () => {
    setIsOpen(false);
  };

  const handlePaste = (e: React.ClipboardEvent<HTMLInputElement>) => {
    const clipboardData = e.clipboardData;
    if (clipboardData) {
      const pastedData = clipboardData.getData('text');
      apiKeyStorage.set(pastedData);
    }
  };

  const handleInputClick = () => {
    if (inputRef.current) {
      inputRef.current.select();
    }
  };

  const handleCopyKey = () => {
    navigator.clipboard.writeText(apiKey);
  };

  const handleDelete = () => {
    apiKeyStorage.set('');
  };

  return (
    <div className="space-y-2">
      <h3 className="font-medium text-slate-800 dark:text-slate-200">{API_KEY_SETTING}</h3>
      {!apiKey && (
        <div className="text-red-500 flex items-center text-xs">
          <AlertTriangle className="h-3 w-3 mr-1" />
          {PLEASE_SET_API_KEY}
        </div>
      )}
      <div className="relative">
        <Input
          ref={inputRef}
          type={showKey ? 'text' : 'password'}
          placeholder="OpenAI API Key"
          value={apiKey}
          readOnly
          onPaste={handlePaste}
          onClick={handleInputClick}
          className="pr-20 bg-slate-100 dark:bg-slate-800 text-slate-800 dark:text-slate-200 placeholder-slate-500 dark:placeholder-slate-400 cursor-pointer"
        />
        <div className="absolute right-2 top-1/2 -translate-y-1/2 flex items-center gap-1">
          <Button
            variant="ghost"
            size="icon"
            data-testid="eye-icon"
            className="h-8 w-8 text-slate-500 hover:text-slate-900"
            onClick={() => setShowKey(!showKey)}
          >
            {showKey ? <EyeOff className="h-4 w-4" /> : <Eye className="h-4 w-4" />}
          </Button>
          {apiKey && <CopyButton handleCopyText={handleCopyKey} />}
        </div>
      </div>
      <Select value={apiVersion} onValueChange={apiVersionStorage.set}>
        <SelectTrigger className="w-full bg-slate-100 dark:bg-slate-800 text-slate-800 dark:text-slate-200">
          <SelectValue placeholder="Select API Version" />
        </SelectTrigger>
        <SelectContent>
          <SelectItem value="gpt-4o-mini">GPT-4o-mini</SelectItem>
          <SelectItem value="gpt-4o">GPT-4o</SelectItem>
          <SelectItem value="o1-mini">GPT-o1-mini</SelectItem>
          <SelectItem value="o1">GPT-o1</SelectItem>
        </SelectContent>
      </Select>
      <div className="flex gap-2">
        <Button
          disabled={!apiKey}
          data-testid="save-icon"
          className="flex-1 bg-emerald-500 text-white hover:bg-emerald-600 disabled:bg-slate-300 dark:disabled:bg-slate-700"
          onClick={handleClosePopover}
        >
          {SAVE}
        </Button>
        {apiKey && (
          <Button data-testid="delete-api-icon" variant="destructive" size="icon" onClick={handleDelete}>
            <Trash2 className="h-4 w-4" />
          </Button>
        )}
      </div>
    </div>
  );
};
