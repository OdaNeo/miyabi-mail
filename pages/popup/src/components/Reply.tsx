import type { PROMPT_KEYS } from '@src/utils/tts';
import { AnimatePresence, motion } from 'framer-motion';
import { Label } from './ui/label';
import { ChevronDown, ChevronUp, ClipboardCopy } from 'lucide-react';
import { Button } from './ui/button';
import { cn } from '@src/lib/utils';
import { useState } from 'react';
import { MessageBox } from './ui/MessageBox';
import { replyStorage, translationStorage } from '@extension/storage';
import { useStorage } from '@extension/shared';
import { useGeneratedStore } from '@src/store/generatedStore';
import { useExpandedSectionStore } from '@src/store/expandedSectionStore';

export function Reply() {
  const reply = useStorage(replyStorage);
  const translation = useStorage(translationStorage);

  const { isGenerated } = useGeneratedStore();
  const { expandedSection, setExpandedSection } = useExpandedSectionStore();

  const [copied, setCopied] = useState(false);

  const toggleSection = (section: PROMPT_KEYS) => {
    setExpandedSection(expandedSection === section ? null : section);
  };

  const handleCopy = () => {
    navigator.clipboard
      .writeText(reply)
      .then(() => {
        setCopied(true);
        setTimeout(() => setCopied(false), 2000);
      })
      .catch(err => {
        console.error('Failed to copy text: ', err);
      });
  };

  return (
    <AnimatePresence>
      {isGenerated && (
        <motion.div
          initial={{ opacity: 0, height: 0 }}
          animate={{ opacity: 1, height: 'auto' }}
          exit={{ opacity: 0, height: 0 }}
          className="space-y-2 overflow-hidden"
        >
          {(['TRANSLATION', 'REPLY'] as PROMPT_KEYS[]).map(section => (
            <div key={section} className="space-y-1">
              <div
                className="flex justify-between items-center cursor-pointer p-1 rounded hover:bg-slate-100 dark:hover:bg-slate-800 transition-colors"
                onClick={() => toggleSection(section)}
                onKeyUp={e => {
                  if (e.key === 'Enter' || e.key === ' ') {
                    toggleSection(section);
                  }
                }}
                role="button"
                tabIndex={0}
              >
                <Label className="text-xs font-medium text-slate-700 dark:text-slate-300">
                  {section === 'TRANSLATION' ? '原文の翻訳：' : '日本語の返信：'}
                </Label>
                <div className="flex items-center space-x-1">
                  {section === 'REPLY' && (
                    <Button
                      variant="outline"
                      size="sm"
                      onClick={e => {
                        e.stopPropagation();
                        handleCopy();
                      }}
                      className={cn(
                        'h-6 text-xs transition-colors',
                        copied
                          ? 'bg-emerald-500 text-white hover:bg-emerald-600'
                          : 'text-slate-600 dark:text-slate-400',
                      )}
                    >
                      {copied ? 'コピー済み' : 'コピー'}
                      <ClipboardCopy className="h-3 w-3 ml-1" />
                    </Button>
                  )}
                  {expandedSection === section ? (
                    <ChevronUp className="h-3 w-3 text-slate-500 dark:text-slate-400" />
                  ) : (
                    <ChevronDown className="h-3 w-3 text-slate-500 dark:text-slate-400" />
                  )}
                </div>
              </div>
              {expandedSection === section && (
                <MessageBox>{section === 'TRANSLATION' ? translation : reply}</MessageBox>
              )}
            </div>
          ))}
        </motion.div>
      )}
    </AnimatePresence>
  );
}
