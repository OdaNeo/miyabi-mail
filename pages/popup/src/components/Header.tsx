import { Tooltip, TooltipContent, TooltipProvider, TooltipTrigger } from '@src/components/ui/tooltip';
import { Button } from '@src/components/ui/button';
import { Settings, Sun, Moon, Globe } from 'lucide-react';
import { Popover, PopoverContent, PopoverTrigger } from '@src/components/ui/popover';
import { useStorage } from '@extension/shared';
import { darkModeStorage, i18nStorage } from '@extension/storage';
import { useOpenStore } from '@src/store/openStore';
import { useI18n } from '@src/hooks/useI18n';
import { SettingPopoverContent } from './SettingPopoverContent';

export function Header() {
  const darkMode = useStorage(darkModeStorage);

  const { isOpen, setIsOpen } = useOpenStore();

  const { CHANGE_TO_LIGHT, CHANGE_TO_DARK, CHANGE_LANGUAGE } = useI18n();

  const handleSwitchLanguage = () => {
    i18nStorage.next();
  };

  const handleToggleDarkMode = () => {
    document.documentElement.classList.toggle('dark');
    darkModeStorage.set(!darkMode);
  };

  const handleTogglePopover = () => {
    setIsOpen(!isOpen);
  };

  return (
    <div className="flex justify-between items-center">
      <h2 className="text-base font-bold text-slate-800 dark:text-slate-100">雅返信</h2>
      <div className="flex space-x-1">
        <TooltipProvider>
          <Tooltip>
            <TooltipTrigger asChild>
              <Button
                variant="outline"
                size="icon"
                className="w-6 h-6 text-slate-600 dark:text-slate-400 hover:bg-slate-200 dark:hover:bg-slate-700"
                onClick={handleSwitchLanguage}
              >
                <Globe data-testid="globe-icon" className="h-3 w-3" />
              </Button>
            </TooltipTrigger>
            <TooltipContent className="bg-slate-700 text-slate-100 dark:bg-slate-200 dark:text-slate-800">
              {CHANGE_LANGUAGE}
            </TooltipContent>
          </Tooltip>
        </TooltipProvider>
        <TooltipProvider>
          <Tooltip>
            <TooltipTrigger asChild>
              <Button
                variant="outline"
                size="icon"
                data-testid="toggle-dark-mode"
                className="w-6 h-6 text-slate-600 dark:text-slate-400 hover:bg-slate-200 dark:hover:bg-slate-700"
                onClick={handleToggleDarkMode}
              >
                {darkMode ? <Sun className="h-3 w-3" /> : <Moon className="h-3 w-3" />}
              </Button>
            </TooltipTrigger>
            <TooltipContent className="bg-slate-700 text-slate-100 dark:bg-slate-200 dark:text-slate-800">
              <p>{darkMode ? CHANGE_TO_LIGHT : CHANGE_TO_DARK}</p>
            </TooltipContent>
          </Tooltip>
        </TooltipProvider>
        <Popover open={isOpen} onOpenChange={setIsOpen}>
          <PopoverTrigger asChild>
            <Button
              onClick={handleTogglePopover}
              variant="outline"
              size="icon"
              data-testid="setting-icon"
              className="w-6 h-6 text-slate-600 dark:text-slate-400 hover:bg-slate-200 dark:hover:bg-slate-700"
            >
              <Settings className="h-3 w-3" />
            </Button>
          </PopoverTrigger>
          <PopoverContent className="w-60 bg-white dark:bg-slate-900 border border-slate-200 dark:border-slate-700">
            <SettingPopoverContent setIsOpen={setIsOpen} />
          </PopoverContent>
        </Popover>
      </div>
    </div>
  );
}
