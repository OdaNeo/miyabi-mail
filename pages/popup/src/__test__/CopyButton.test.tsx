import { describe, it, expect, vi, beforeEach, afterEach } from 'vitest';
import { render, screen, fireEvent, act } from '@testing-library/react';
import { CopyButton } from '../components/CopyButton';

vi.mock('framer-motion', () => ({
  motion: {
    div: ({ children, ...props }: any) => <div {...props}>{children}</div>,
  },
  AnimatePresence: ({ children }: any) => <>{children}</>,
}));

vi.mock('lucide-react', () => ({
  Copy: () => <div data-testid="copy-icon">Copy</div>,
  Check: () => <div data-testid="check-icon">Check</div>,
}));

describe('CopyButton', () => {
  const mockHandleCopyText = vi.fn();

  beforeEach(() => {
    vi.clearAllMocks();
    vi.useFakeTimers();
  });

  afterEach(() => {
    vi.runOnlyPendingTimers();
    vi.useRealTimers();
  });

  it('should match snapshot in initial state', () => {
    const { container } = render(<CopyButton handleCopyText={mockHandleCopyText} />);
    expect(container).toMatchSnapshot();
  });

  it('should match snapshot in copied state', async () => {
    const { container } = render(<CopyButton handleCopyText={mockHandleCopyText} />);
    const button = screen.getByRole('button');
    fireEvent.click(button);
    expect(container).toMatchSnapshot();
  });

  it('should call handleCopyText when clicked', () => {
    render(<CopyButton handleCopyText={mockHandleCopyText} />);
    const button = screen.getByRole('button');
    fireEvent.click(button);
    expect(mockHandleCopyText).toHaveBeenCalledTimes(1);
  });

  it('should show check icon after click and revert back after 2 seconds', () => {
    render(<CopyButton handleCopyText={mockHandleCopyText} />);
    const button = screen.getByRole('button');

    expect(screen.getByTestId('copy-icon')).toBeInTheDocument();

    fireEvent.click(button);
    expect(screen.getByTestId('check-icon')).toBeInTheDocument();

    act(() => {
      vi.advanceTimersByTime(2000);
    });
    expect(screen.getByTestId('copy-icon')).toBeInTheDocument();
  });

  it('should not trigger copy again while in copied state', () => {
    render(<CopyButton handleCopyText={mockHandleCopyText} />);
    const button = screen.getByRole('button');

    fireEvent.click(button);
    expect(mockHandleCopyText).toHaveBeenCalledTimes(1);

    fireEvent.click(button);
    expect(mockHandleCopyText).toHaveBeenCalledTimes(1);
  });
});
