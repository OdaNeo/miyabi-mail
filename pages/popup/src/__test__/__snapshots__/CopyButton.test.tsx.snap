// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CopyButton > should match snapshot in copied state 1`] = `
<div>
  <button
    class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 hover:bg-accent h-6 w-6 text-slate-500 hover:text-slate-900 relative"
    data-testid="copy-icon-group"
  >
    <div
      animate="[object Object]"
      class="absolute inset-0 flex items-center justify-center"
      exit="[object Object]"
      initial="[object Object]"
      transition="[object Object]"
    >
      <div
        data-testid="check-icon"
      >
        Check
      </div>
    </div>
  </button>
</div>
`;

exports[`CopyButton > should match snapshot in initial state 1`] = `
<div>
  <button
    class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 hover:bg-accent h-6 w-6 text-slate-500 hover:text-slate-900 relative"
    data-testid="copy-icon-group"
  >
    <div
      animate="[object Object]"
      class="absolute inset-0 flex items-center justify-center"
      exit="[object Object]"
      initial="[object Object]"
      transition="[object Object]"
    >
      <div
        data-testid="copy-icon"
      >
        Copy
      </div>
    </div>
  </button>
</div>
`;
